pkg/userdata/centos/userdata.go

```diff
@@ -0,0 +1,246 @@
+package centos
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"net"
+	"text/template"
+
+	"github.com/Masterminds/semver"
+	"github.com/kubermatic/machine-controller/pkg/containerruntime"
+	machinesv1alpha1 "github.com/kubermatic/machine-controller/pkg/machines/v1alpha1"
+	"github.com/kubermatic/machine-controller/pkg/providerconfig"
+	machinetemplate "github.com/kubermatic/machine-controller/pkg/template"
+	"github.com/kubermatic/machine-controller/pkg/userdata/cloud"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+type Provider struct{}
+
+type Config struct {
+	DistUpgradeOnBoot bool `json:"distUpgradeOnBoot"`
+}
+
+type packageCompatibilityMatrix struct {
+	versions []string
+	pkg      string
+}
+
+var dockerInstallCandidates = []packageCompatibilityMatrix{
+	{
+		versions: []string{"1.12", "1.12.6"},
+		pkg:      "docker-1.12.6",
+	},
+	{
+		versions: []string{"1.13", "1.13.1"},
+		pkg:      "docker-1.13.1",
+	},
+}
+
+func (p Provider) SupportedContainerRuntimes() (runtimes []machinesv1alpha1.ContainerRuntimeInfo) {
+	for _, installCandidate := range dockerInstallCandidates {
+		for _, v := range installCandidate.versions {
+			runtimes = append(runtimes, machinesv1alpha1.ContainerRuntimeInfo{Name: containerruntime.Docker, Version: v})
+		}
+	}
+	return runtimes
+}
+
+func getConfig(r runtime.RawExtension) (*Config, error) {
+	p := Config{}
+	if len(r.Raw) == 0 {
+		return &p, nil
+	}
+	if err := json.Unmarshal(r.Raw, &p); err != nil {
+		return nil, err
+	}
+	return &p, nil
+}
+
+func getDockerPackageName(version string) (string, error) {
+	for _, installCandidate := range dockerInstallCandidates {
+		for _, v := range installCandidate.versions {
+			if v == version {
+				return installCandidate.pkg, nil
+			}
+		}
+	}
+	return "", fmt.Errorf("no package found for version '%s'", version)
+}
+
+func (p Provider) UserData(spec machinesv1alpha1.MachineSpec, kubeconfig string, ccProvider cloud.ConfigProvider, clusterDNSIPs []net.IP, kubernetesCACert string) (string, error) {
+	tmpl, err := template.New("user-data").Funcs(machinetemplate.TxtFuncMap()).Parse(ctTemplate)
+	if err != nil {
+		return "", fmt.Errorf("failed to parse user-data template: %v", err)
+	}
+
+	semverKubeletVersion, err := semver.NewVersion(spec.Versions.Kubelet)
+	if err != nil {
+		return "", fmt.Errorf("invalid kubelet version: '%v'", err)
+	}
+	kubeletVersion := fmt.Sprintf("-%s", semverKubeletVersion.String())
```

Comment:
Move the `-`  to the template. It's not obvious why it should be added here already.
Otherwise it will create confusion when this variable will be used in some other context